spec/rubocop/cop/rails/timecop_spec.rb

```diff
@@ -0,0 +1,177 @@
+# frozen_string_literal: true
+
+RSpec.describe(RuboCop::Cop::Rails::Timecop, :config) do
+  subject(:cop) { described_class.new(config) }
+
+  describe 'Timecop.freeze' do
+    context 'without a block' do
+      context 'without arguments' do
+        it 'adds an offense' do
+          expect_offense(<<-RUBY.strip_indent)
+            Timecop.freeze
+            ^^^^^^^^^^^^^^ Use `freeze_time` instead of `Timecop.freeze`
+          RUBY
+        end
+
+        it 'autocorrects to `freeze_time`' do
+          expect(autocorrect_source('Timecop.freeze')).to(eq('freeze_time'))
+        end
+
+        context 'spread over multiple lines' do
+          it 'adds an offense' do
+            expect_offense(<<-RUBY.strip_indent)
+              Timecop
+              ^^^^^^^ Use `freeze_time` instead of `Timecop.freeze`
+                .freeze
+            RUBY
+          end
+
+          it 'autocorrects to `freeze_time`' do
+            expect(autocorrect_source("Timecop\n  .freeze"))
+              .to(eq('freeze_time'))
+          end
+        end
+      end
+
+      context 'with arguments' do
+        it 'adds an offense' do
+          expect_offense(<<-RUBY.strip_indent)
+            Timecop.freeze(123)
+            ^^^^^^^^^^^^^^^^^^^ Use `travel` or `travel_to` instead of `Timecop.freeze`
+          RUBY
+        end
+
+        it 'does not autocorrect' do
+          source = 'Timecop.freeze(123)'
+
+          expect(autocorrect_source(source)).to(eq(source))
+        end
+      end
+    end
+
+    context 'with a block' do
+      context 'without arguments' do
+        it 'adds an offense' do
+          expect_offense(<<-RUBY.strip_indent)
+            Timecop.freeze { }
+            ^^^^^^^^^^^^^^ Use `freeze_time` instead of `Timecop.freeze`
+          RUBY
+        end
+
+        it 'autocorrects to `freeze_time`' do
+          expect(autocorrect_source('Timecop.freeze { }'))
+            .to(eq('freeze_time { }'))
+        end
+      end
+
+      context 'with arguments' do
+        it 'adds an offense' do
+          expect_offense(<<-RUBY.strip_indent)
+            Timecop.freeze(123) { }
+            ^^^^^^^^^^^^^^^^^^^ Use `travel` or `travel_to` instead of `Timecop.freeze`
+          RUBY
+        end
+
+        it 'does not autocorrect' do
+          source = 'Timecop.freeze(123) { }'
+
+          expect(autocorrect_source(source)).to(eq(source))
+        end
+      end
+    end
+  end
+
+  describe 'Timecop.return' do
+    context 'without a block' do
+      it 'adds an offense' do
+        expect_offense(<<-RUBY.strip_indent)
+          Timecop.return
+          ^^^^^^^^^^^^^^ Use `travel_back` instead of `Timecop.return`
+        RUBY
+      end
+
+      context 'in Rails < 6.0', :rails5 do
```

Comment:
It seems that `TimeHelpers` were introduced in Rails 4.1, and by that time `freeze_time` was not yet available until Rails 4.2.
Unfortunately, RuboCop does not provide minor version, so we're left with two options:
 - raise offences for Rails 4.0, including for `freeze` for Rails 4.1, which is unsafe;
 - skip all the detection for Rails prior to 5.0

Reply:
Looks like you have a version error here, though your point is still valid.

- **4.1** introduces `TimeHelpers#travel`, `TimeHelpers#travel_to`, & `TimeHelpers#travel_back`
- **5.2** introduces `TimeHelpers#freeze_time` alias for `TimeHelpers#travel_to(Time.now)`
- **6.0** _will introduce_ `TimeHelpers#unfreeze_time` alias for `TimeHelpers#travel_back`

Given `freeze_time` is [implemented](https://github.com/rails/rails/blob/94b5cd3a20edadd6f6b8cf0bdf1a4d4919df86cb/activesupport/lib/active_support/testing/time_helpers.rb#L189-L191) by passing `Time.now` to `travel_to`, every mention of `freeze_time` is safely replaced with `travel_to(Time.now)`.

----

I'm surprised about the lack of minor version specification, as that means the experience we can provide is rather poor.

Ideally, we would provide the following experience:

- **Version < 4.1**: No offences or corrections
- **4.1 >= Versions < 5.2**: Add offences & corrections, using only `travel` & `travel_to`
- **5.2 >= Versions < 6.0**: Change offences & corrections to include `freeze_time`
- **6.0 >= Versions**: Change offences & corrections to include `unfreeze_time`

With major version restrictions, that means we can only provide the following:

- **Versions < 5.0**: No offences or corrections
- **5.0 >= Versions < 6.0**: Add offences & corrections, using only `travel` & `travel_to`
- **6.0 >= Versions**: Change offences & corrections to include `freeze_time` & `unfreeze_time`

----

Can you point me in the right direction if I want to investigate version granularity? I did try to look into where RSpec gets the various `:rails4`/`:rails5` contexts, but I couldn't figure it out. We mostly use Minitest style tests at Shopify 😅

Reply:
It turns out it's possible to specify a more granular Rails version in specs, e.g. by defining it as         
```
let(:rails_version) { '5.1.4' }
```
as it's possible to specify it as such in `.rubocop.yml` file.
I'm not sure if it's worth adding those as shared contexts, but this can be used in the cop specs, and in cop itself (I imagine something like a per-Rails-version strategy).
WDYT?

Sorry for late response. Are you still up for moving this forward?

Reply:
Yup 👍 I'll swing back around to this when I have time.